les, light: implement ODR transaction lookup by hash

zsfelfoldi opened the pull request:
This PR implements an ODR request for retrieving transaction status and adds it to the API backend so that it is used by GetTransactionByHash and GetRawTransactionByHash. From the user perspective this means that eth.getTransaction and eth.getRawTransaction will now work in light mode too.
Solves https://github.com/ethereum/go-ethereum/issues/19063

rjl493456442 commented on les/handler.go:
```diff
@@ -1119,14 +1119,21 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
+			MsgType: MsgTxStatus,
+			ReqID:   resp.ReqID,
+			Obj:     resp.Status,
+		}
```
`SendTxV2Msg` will also respond a `TxStatusMsg`. Unfortunately the `SendTxV2Msg` is not sent by `retriever`. So that the delivered message will cause a peer error.

zsfelfoldi replied:
Thanks for finding this one, I missed it because when I wrote this code the tx relay still used the old LES/1 SendTx which did not receive a reply. Fixed, now the tx relay uses the retriever too, with an empty validation function so it does not care about the contents of the reply.
Btw the reply could be used meaningfully to report when the tx is not accepted to the server's tx pool but that requires rewriting the light tx pool which should happen in a later release. Now the operation is identical to the way it has worked until now.

zsfelfoldi commented on les/odr_requests.go:
```diff
@@ -557,6 +559,44 @@ func (r *BloomRequest) Validate(db ethdb.Database, msg *Msg) error {
+		return errInvalidMessageType
+	}
+	status := msg.Obj.([]light.TxStatus)
+	if len(status) != 1 {
```
done

zsfelfoldi commented on light/odr.go:
```diff
@@ -177,3 +177,20 @@ func (req *BloomRequest) StoreResult(db ethdb.Database) {
+}
+
+// TxRequest is the ODR request type for retrieving transaction status
+type TxRequest struct {
```
done

holiman commented on eth/api_backend.go:
```diff
@@ -174,6 +174,10 @@ func (b *EthAPIBackend) GetPoolTransaction(hash common.Hash) *types.Transaction
 }
 
+func (b *EthAPIBackend) GetTransactionWithOdr(ctx context.Context, txHash common.Hash) (*types.Transaction, common.Hash, uint64, uint64, error) {
+	return nil, common.Hash{}, 0, 0, nil
```
Why not return a `not implemented` - error here?

karalabe commented on internal/ethapi/backend.go:
```diff
@@ -62,6 +62,7 @@ type Backend interface {
 
 	// TxPool API
 	SendTx(ctx context.Context, signedTx *types.Transaction) error
+	GetTransactionWithOdr(ctx context.Context, txHash common.Hash) (*types.Transaction, common.Hash, uint64, uint64, error)
```
Pls rename to `GetTransaction` and leave the ODR decision handling internally.

holiman commented:
PR triage summary: the user/caller doesn't care about where the transaction comes from, he just wants a transaction. So it would be better not to expose `GetTransaction`/`GetTransactionODR` , but instead just `GetTransaction` and the node should figure it out internally whether it exists locally or should be fetched from the network

zsfelfoldi commented:
@holiman @karalabe @rjl493456442 I added a commit that renames GetTransactionWithOdr to GetCanonicalTransaction which is now also implemented in eth/api_backend.go. I also made the error handling consistent between GetTransactionByHash and GetRawTransactionByHash (now both of them return an error if the retrieval from the backend has failed and we have no information on whether the given transaction is in the chain or not).

zsfelfoldi commented:
It will "just work". No extra function is provided, `eth.getTransaction` works in light mode.

karalabe commented on eth/api_backend.go:
```diff
@@ -174,6 +175,11 @@ func (b *EthAPIBackend) GetPoolTransaction(hash common.Hash) *types.Transaction
 	return b.eth.txPool.Get(hash)
 }
 
+func (b *EthAPIBackend) GetCanonicalTransaction(ctx context.Context, txHash common.Hash) (*types.Transaction, common.Hash, uint64, uint64, error) {
```
GetCanonicalTransaction -> GetTransaction

karalabe commented on light/odr_util.go:
```diff
@@ -227,3 +228,18 @@ func GetBloomBits(ctx context.Context, odr OdrBackend, bitIdx uint, sectionIdxLi
 	}
 }
+
+// GetTransaction retrieves a canonical transaction by hash and also returns its position in the chain
```
Why does this dangle in the air? Why isn't this attached to a light chain or something?

zsfelfoldi replied:
It is attached to the `OdrBackend` which is attached to the correct chain. Just like direct database read functions where the chain is selected by the supplied chain database.
It wasn't correct though because I forgot to check whether the header is available and canonical. It is added now.
